File: seer/src/one_resolver.rs
use std::collections::{HashMap, HashSet};
use std::fmt::Debug;
use std::hash::Hash;

#[derive(Clone)]
pub struct OneResolver<T: Eq + Hash + Ord + Clone + Debug, F: Eq + Hash + Ord + Clone + Debug> {
    tables_by_field: HashMap<F, HashSet<T>>,
    candidates: HashSet<T>,
    fields: Vec<F>,
}

impl<T: Eq + Hash + Ord + Clone + Debug, F: Eq + Hash + Ord + Clone + Debug> OneResolver<T, F> {
    pub fn new(tables: HashMap<T, Vec<F>>) -> Self {
        let mut tables_by_field: HashMap<F, HashSet<T>> = HashMap::new();
        let mut candidates = HashSet::new();

        for (t, fields) in &tables {
            candidates.insert(t.clone());

            for field in fields {
                tables_by_field
                    .entry(field.clone())
                    .or_default()
                    .insert(t.clone());
            }
        }

        let mut resolver = Self {
            tables_by_field,
            candidates,
            fields: Vec::new(),
        };

        resolver.refresh();

        resolver
    }

    fn refresh(&mut self) {
        let fields: Vec<F> = self.tables_by_field
            .iter()
            .filter(|(_, tables)| tables.iter().any(|t| self.candidates.contains(t)))
            .map(|(field, _)| field.clone())
            .collect();

        self.fields = fields;
    }

    pub fn get_fields(&self) -> &[F] {
        &self.fields
    }

    pub fn use_field(&mut self, field: F) -> Option<bool> {
        let tables = self.tables_by_field.get(&field)?;
        let next: HashSet<T> = self.candidates.intersection(tables).cloned().collect();

        if next.is_empty() {
            return Some(false);
        }

        self.candidates = next;

        self.refresh();

        Some(true)
    }

    pub fn use_table(&mut self, table: &T) -> Option<()> {
        if !self.candidates.contains(table) {
            return None;
        }

        self.candidates.clear();

        Some(())
    }

    pub fn get_required_tables(&self) -> Vec<T> {
        self.candidates.iter().cloned().collect()
    }

    pub fn is_satisfied(&self) -> bool {
        self.candidates.len() == 0
    }
}

File: experiments/9-advanced-modelling/one_resolver.py
class OneResolver:
    def __init__(self, tables: dict[str, list[str]]):
        self.tables_by_field: dict[str, set[str]] = {}
        self.candidates: set[str] = set()
        self.tables: set[str] = set(tables.keys())

        for table, fields in tables.items():
            self.candidates.add(table)
            
            for field in fields:
                self.tables_by_field.setdefault(field, set()).add(table)

        self.fields: set[str] = set()

        self.refresh_fields()

    def refresh_fields(self):
        self.fields = {
            f for f, tables in self.tables_by_field.items()
            if tables & self.candidates
        }

    def get_fields(self) -> set[str]:
        return self.fields

    def use_field(self, field: str):
        tables = self.tables_by_field.get(field)

        if tables is None:
            raise Exception(f"Field {field} does not exists")

        next_candidates = self.candidates & tables

        if not next_candidates:
            raise Exception(f"No intersection between candidates & tables of the field")

        self.candidates = next_candidates

        self.refresh_fields()

    def use_table(self, name: str):
        if name not in self.candidates:
            raise Exception(f"Table is not a candidate")
        
        self.candidates.clear()

    def get_required_tables(self) -> set[str]:
        return self.candidates

    def get_excluded_tables(self) -> set[str]:
        return self.tables - self.candidates

    def is_satisfied(self) -> bool:
        return len(self.candidates) == 0
